### PR TITLE
Documentation

### PR DESCRIPTION
**Pull request by bitpshr**
This PR adds JSDoc where applicable to the following modules:

```
app/scripts/config.js
app/scripts/contentscript.js
app/scripts/edge-encryptor.js
app/scripts/first-time-state.js
app/scripts/inpage.js
app/scripts/popup-core.js
app/scripts/controllers/computed-balances.js
app/scripts/lib/createProviderMiddleware.js
app/scripts/lib/port-stream.js
app/scripts/lib/setupMetamaskMeshMetrics.js
```

Refs #3941

### CODE REVIEW

**Review by danfinlay — CHANGES_REQUESTED** [review]
A few spelling changes, a conceptual note, a few fun facts, and maybe we remove the jsdoc build output from git?

File: app/scripts/popup-core.js [hunk 1]
```diff
@@ -7,10 +7,14 @@ const launchMetamaskUi = require('../../ui')
...
 const setupMultiplex = require('./lib/stream-utils.js').setupMultiplex
 
-
 module.exports = initializePopup
 
-
+/**
+ * Asynchronously initializes the MetaMask popup UI
+ *
+ * @param {{ container: Element, connectionStream: any }} config Popup configuration object
```
[the remark this answers is not in the document]
Comment (bitpshr) [reply app/scripts/popup-core.js]: JSDoc technically supports most DOM interfaces as actual types. `Element`, or even `HTMLElement` to be more specific, are valid JSDoc types. `any` is actually TypeScript, so good catch there. It should be an asterisk.

File: app/scripts/popup-core.js [hunk 2]
```diff
@@ -28,6 +38,11 @@ function connectToAccountManager (connectionStream, cb) {
   setupWeb3Connection(mx.createStream('provider'))
 }
 
+/**
+ * Establishes a streamed connection to a Web3 provider
+ *
+ * @param {any} connectionStream PortStream instance establishing a background connection
```
[the remark this answers is not in the document]
Comment (bitpshr) [reply app/scripts/popup-core.js]: Agreed. We could define one ourselves, but TypeScript...built in...soon 🙏

File: docs/jsdocs/config.js.html [hunk 3]
```diff
@@ -0,0 +1,148 @@
+<!DOCTYPE html>
```
[the remark this answers is not in the document]
Comment (bitpshr) [reply docs/jsdocs/config.js.html]: Cool. For PR clarity I removed them from version control as part of this PR and updated the `.gitignore`.

File: .gitignore [hunk 4]
```diff
@@ -20,6 +20,7 @@ dist
 builds/
 disc/
 builds.zip
+docs
```
Comment (danfinlay) [inline .gitignore]: Only ignore `docs/jsdocs` please! Other docs are good, just trying to reduce things that can be rebuilt on each commit.

**Review by danfinlay — CHANGES_REQUESTED** [review]
Sorry, I only wanted `jsdocs` out of repo, other docs are good!

**Comment by bitpshr** [comment]
@danfinlay good catch, thanks. PR updated.

**Comment by bitpshr** [comment]
A few more files documented @danfinlay.

File: app/scripts/contentscript.js [hunk 5]
```diff
@@ -89,17 +96,34 @@ function setupStreams () {
...
 function logStreamDisconnectWarning (remoteLabel, err) {
   let warningMsg = `MetamaskContentscript - lost connection to ${remoteLabel}`
   if (err) warningMsg += '\n' + err.stack
   console.warn(warningMsg)
 }
 
+/**
+ * Determines if Web3 should be injected
+ *
+ * @returns {boolean} True of Web3 should be injected
```
[the remark this answers is not in the document]
Comment (bitpshr) [reply app/scripts/contentscript.js]: I didn't know this was actually valid for type in JSDoc. I thought this was only a shortcut to render `true` (or any other code fragment) in place of a parameter. I'm inclined to leave `boolean` to let parsers do their thing.

[the remark this answers is not in the document]
Comment (whymarrh) [reply app/scripts/contentscript.js]: Sorry, yeah, I was referring to the description: `{@code true} of Web3 should be injected`

[the remark this answers is not in the document]
Comment (whymarrh) [reply app/scripts/contentscript.js]: (Also just noticed "of" should be "if".)

File: app/scripts/popup-core.js [hunk 6]
```diff
@@ -19,6 +23,12 @@ function initializePopup ({ container, connectionStream }, cb) {
   ], cb)
 }
 
+/**
+ * Establishes streamed connections to background scripts and a Web3 provider
+ *
+ * @param {*} connectionStream PortStream instance establishing a background connection
```
[the remark this answers is not in the document]
Comment (bitpshr) [reply app/scripts/popup-core.js]: We should actually just be able to use `PortDuplexStream` as a type. Updated.

File: app/scripts/contentscript.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bitpshr) [reply app/scripts/contentscript.js]: Ah, I see. Good call!

**Comment by bitpshr** [comment]
@whymarrh thanks for the great feedback. This should be ready for another review when you have a chance.

**Review by whymarrh — APPROVED** [review]
LGTM!

**Comment by bitpshr** [comment]
@danfinlay this should be ready for another review when you have a chance. Thanks!

**Review by danfinlay — APPROVED** [review]
Awesome, thank you!

**Pull request merged**